bpfd/src/server/bpf.rs

```diff
@@ -216,89 +252,234 @@ impl<'a> BpfManager<'a> {
+
+                self.attach_extensions_tc_in(&if_index, &mut dispatcher_loader, Some(ext_loader))?;
+
+                self.attach_or_replace_dispatcher_tc_in(
+                    iface.clone(),
+                    if_index,
+                    dispatcher_loader,
+                )?;
+
+                if let Some(r) = old_revision {
+                    self.cleanup_extensions(if_index, r)?;
+                }
+
+                info!(
+                    "Program added: {} programs attached to {}",
+                    self.progs_tc_in.get(&if_index).unwrap().len(),
+                    &iface,
+                );
+                Ok(id)
+            }
+            2 => Err(BpfdError::UnsuportedProgramType),
+            _ => Err(BpfdError::UnsuportedProgramType),
         }
+    }
 
-        // Calculate the dispatcher revision
-        let (old_revision, revision) = if let Some(old_revision) = self.revisions.remove(&if_index)
-        {
-            let r = old_revision.wrapping_add(1);
-            self.revisions.insert(if_index, r);
-            (Some(old_revision), r)
-        } else {
-            self.revisions.insert(if_index, 0);
-            (None, 0)
-        };
+    // ANF-TODO: Fix this to work for multiple program types.
+    // ANF-TODO: User may need to pass in the program type to remove, or we may need to
+    //           search all "progs" hashmaps for the program.
+    pub(crate) fn remove_program(
+        &mut self,
+        id: Uuid,
+        iface: String,
+        owner: String,
+    ) -> Result<(), BpfdError> {
+        let if_index = self.get_ifindex(&iface)?;
+        if let Some(programs) = self.progs_tc_in.get_mut(&if_index) {
+            if let Some(prog) = programs.get(&id) {
+                if !(prog.owner == owner || owner == SUPERUSER) {
+                    return Err(BpfdError::NotAuthorized);
+                }
+            } else {
+                return Err(BpfdError::InvalidID);
+            }
 
-        let prog = ExtensionProgram {
-            path,
-            current_position: None,
-            metadata: Metadata {
-                priority,
-                name: section_name,
-                attached: false,
-            },
-            owner,
-            proceed_on,
-        };
-        prog.save(id)
-            .map_err(|_| BpfdError::Error("unable to persist program data".to_string()))?;
+            let prog = programs.remove(&id).unwrap();
+            prog.delete(id)
+                .map_err(|_| BpfdError::Error("unable to delete program data".to_string()))?;
+
+            let old_revision = self.revisions_tc_in.remove(&if_index).unwrap();
 
-        self.programs.get_mut(&if_index).unwrap().insert(id, prog);
-        self.sort_extensions(&if_index);
+            if programs.is_empty() {
+                self.progs_tc_in.remove(&if_index);
+                self.dispatchers_tc_in.remove(&if_index);
+                self.cleanup_extensions(if_index, old_revision)?;
```

Comment:
cleanup_extensions() added here to remove the /var/run/bpfd/fs/dispatcher_{if_index}_{revision} which caused the dispatcher to be unloaded.